airbyte-config/config-persistence/src/main/java/io/airbyte/config/persistence/StatePersistence.java

```
@@ -0,0 +1,280 @@
+    };
+  }
+
+  /**
+   * Create or update the states described in the StateWrapper
+   * Null states will be deleted.
+   *
+   * @param connectionId
+   * @param state
+   * @throws IOException
+   */
+  public void updateOrCreateState(final UUID connectionId, final StateWrapper state) throws IOException {
+    final Optional<StateWrapper> previousState = getCurrentState(connectionId);
+    final boolean isMigration = previousState.isPresent() && previousState.get().getStateType() == StateType.LEGACY &&
+        state.getStateType() != StateType.LEGACY;
+
+    // The only case where we allow a state migration is moving from LEGACY.
+    // We expect any other migration to go through an explicit reset.
+    if (!isMigration && previousState.isPresent() && previousState.get().getStateType() != state.getStateType()) {
+      throw new IOException("Unexpected type migration from '" + previousState.get().getStateType() + "' to '" + state.getStateType() +
+          "'. Migration of StateType need to go through an explicit reset.");
+    }
+
+    this.database.transaction(ctx -> {
+      if (isMigration) {
+        clearLegacyState(ctx, connectionId);
+      }
+      switch (state.getStateType()) {
+        case GLOBAL -> saveGlobalState(ctx, connectionId, state.getGlobal().getGlobal());
+        case STREAM -> saveStreamState(ctx, connectionId, state.getStateMessages());
+        case LEGACY -> saveLegacyState(ctx, connectionId, state.getLegacyState());
+      }
+      return null;
+    });
+  }
+
+  private void clearLegacyState(final DSLContext ctx, final UUID connectionId) {
+    writeStateToDb(ctx, connectionId, null, null, StateType.LEGACY, null);
+  }
+
+  private void saveGlobalState(final DSLContext ctx, final UUID connectionId, final AirbyteGlobalState globalState) {
+    writeStateToDb(ctx, connectionId, null, null, StateType.GLOBAL, globalState.getSharedState());
+    for (final AirbyteStreamState streamState : globalState.getStreamStates()) {
```

Nit:  Any reason not to use a stream/lambda for this?  Are we concerned about swallowing any exceptions raised by `writeStateToDb`?

---

That is very procedural in the end, a for loop felt more straightforward in this case.
About the exceptions, I'd rather let them bubble up for debugging errors if any.